src/test/libcephfs/monaddrsfromcct.cc

```diff
@@ -0,0 +1,103 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "gtest/gtest.h"
+#include "include/cephfs/libcephfs.h"
+#include "common/async/context_pool.h"
+#include "common/ceph_context.h"
+#include "mon/MonClient.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <dirent.h>
+#ifdef __linux__
+#include <sys/xattr.h>
+#endif
+
+// Helper to remove/unset all possible mon information from ConfigProxy
+void helper_clear_cct_of_mon_config(CephContext *cct) {
+  auto& conf = cct->_conf;
+  if (const auto monmap = conf.get_val<std::string>("monmap");
+    !monmap.empty()) {
+    ASSERT_EQ(0, conf.set_val("monmap", "", nullptr));
+  }
+  if (const auto mon_host = conf.get_val<std::string>("mon_host");
+      !mon_host.empty()) {
+    ASSERT_EQ(0, conf.set_val("mon_host", "", nullptr));
+  }
+  if (const auto mon_dns_name = conf.get_val<std::string>("mon_dns_srv_name");
+      !mon_dns_name.empty()) {
+    ASSERT_EQ(0, conf.set_val("mon_dns_srv_name", "", nullptr));
+  }
+}
+
+TEST(LibCephFS, MonaddrsFromCct) {
+  struct ceph_mount_info *ca, *cb;
+  CephContext *cct;
+  ceph::async::io_context_pool icp;
+
+  // Test mount failure when there is no known mon config source
+  ASSERT_EQ(0, ceph_create(&ca, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(ca, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(ca, NULL));
+  cct = ceph_get_mount_context(ca);
+  ASSERT_NE(nullptr, cct);
+  // clear out ConfigProxy
+  helper_clear_cct_of_mon_config(cct);
+  ASSERT_EQ(-ENOENT, ceph_mount(ca, NULL));
+  ceph_shutdown(ca);
+
+  // Test a successful mount with default mon config source in ConfigProxy
+  ASSERT_EQ(0, ceph_create(&ca, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(ca, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(ca, NULL));
+  ASSERT_EQ(0, ceph_mount(ca, NULL));
+
+  // Test a successful mount with only mon values in CephContext, and
+  // no sources in ConfigProxy
+  ASSERT_EQ(0, ceph_create(&cb, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cb, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cb, NULL));
+  cct = ceph_get_mount_context(cb);
+  ASSERT_NE(nullptr, cct);
+
+  // Bootstrap the process of storing mon addrs in CephContext
+  icp.start(cct->_conf.get_val<std::uint64_t>("client_asio_thread_count"));
+  auto mc_bootstrap = new MonClient(cct, icp);
+  ASSERT_EQ(0, mc_bootstrap->get_monmap_and_config());
+  delete mc_bootstrap;
+  icp.stop();
+
+  // Clear all mon config sources in ConfigProxy
+  helper_clear_cct_of_mon_config(cct);
+
+  // Mount should pick up mon config from CephContext and succeed
+  ASSERT_EQ(0, ceph_mount(cb, NULL));
```

Comment:
```suggestion
  cct = ceph_get_mount_context(ca);
  helper_clear_cct_of_mon_config(cct);
  ASSERT_EQ(0, ceph_create_with_context(cb, cct));
```

I think this should work?

Reply:
So the above is what I was going for initially, but post a ceph_mount(**ca**, NULL) the values in ConfigProxy are not modifiable, and return a [EPERM](https://github.com/ceph/ceph/blob/df1a88edcb88c39dbfc4b169deba3c95f4b9142b/src/common/config.cc#L1387-L1397) when I attempt to clear out the same.

`md_config_t::set_safe_to_start_threads` is set to true by, `CephContext::start_service_thread`, which is invoked by `common_init_finish`, which happens in `ceph_mount_info::init` (after it bootstraps the MonClient). This setting prevents the config values from being modified (specifically "mon_host" and "mon_dns_srv_name").

So chasing down the above path, I could not reuse cct from **ca**, hence created a new one.

Now that a new cct was created, I needed to bootstrap it to store the mon addrs in CephContext, which explains the code in lines 77-82, post which I clear the specific mon config before calling ceph_mount(**cb**, NULL).

Having said all this, there are invocations of `md_config_t::_clear_safe_to_start_threads` from [other](https://github.com/ceph/ceph/blob/df1a88edcb88c39dbfc4b169deba3c95f4b9142b/src/test/objectstore/test_bluefs.cc#L66-L72) test [code](https://github.com/ceph/ceph/blob/df1a88edcb88c39dbfc4b169deba3c95f4b9142b/src/test/test_ipaddr.cc#L693-L698), which can hence overcome the above, enabling reuse of cct from **ca**.

I am going to try the above, as it simplifies the test case and its assumptions, assuming this is the more canonical form of doing the same.

Reply:
Ok, so the following works, (and without clearing the safe to start threads flag it does not)

```
  cct = ceph_get_mount_context(ca);
  ASSERT_NE(nullptr, cct);
  cct->_conf._clear_safe_to_start_threads();
  helper_clear_cct_of_mon_config(cct);
  cct->_conf.set_safe_to_start_threads();
  ASSERT_EQ(0, ceph_create_with_context(&cb, cct));
  // Mount should pick up mon config from CephContext and succeed
  ASSERT_EQ(0, ceph_mount(cb, NULL));
```

But, the sequence in this comment has far less assumptions about who sets the flag etc. and I would like to move to this version, if there are no concerns. (this will also not require the lines 77-82)

Reply:
that looks fine. Please leave a comment explaining.

Reply:
Addressed in latest push